File: patterns/maximum-subarray/01.max-sum-subarray-of-size-at-least-k/main.py
class Solution():
    def max_sum_with_k(self, a, n, k):
        sum_k = 0
        
        for i in range(k):
            sum_k += a[i]
            
        j = 0
        last = 0
        ans = sum_k
        
        for i in range(k, n):
            sum_k += a[i]
            last += a[j]
            j += 1
            
            ans = max(ans, sum_k)
            
            if last < 0:
                sum_k -= last
                ans = max(ans, sum_k)
                last = 0
    
        return ans

File: patterns/maximum-subarray/01.max-sum-subarray-of-size-at-least-k/main.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    long long int maxSumWithK(long long int a[], long long int n, long long int k)
    {
        long long int sum = 0;

        for (long long int i = 0; i < k; i++)
        {
            sum += a[i];
        }

        long long int last = 0;
        long long int j = 0;
        long long int ans = LLONG_MIN;
        ans = max(ans, sum);

        for (long long int i = k; i < n; i++)
        {
            sum += a[i];
            last += a[j++];
            ans = max(ans, sum);

            if (last < 0)
            {
                sum -= last;
                ans = max(ans, sum);
                last = 0;
            }
        }

        return ans;
    }
};
